File: test_gigachat_chat.py
import requests
import os
import json
from dotenv import load_dotenv

# Загружаем переменные окружения
load_dotenv('.env.local')

def get_gigachat_token():
    url = "https://ngw.devices.sberbank.ru:9443/api/v2/oauth"
    
    auth_key = os.getenv('GIGACHAT_API_KEY')
    
    if not auth_key:
        print("❌ GIGACHAT_API_KEY не найден")
        return None
    
    payload = {
        'scope': 'GIGACHAT_API_PERS'
    }
    
    headers = {
        'Content-Type': 'application/x-www-form-urlencoded',
        'Accept': 'application/json',
        'RqUID': 'ac160d04-0557-4dc0-bc66-8e7875d69102',
        'Authorization': f'Basic {auth_key}'
    }
    
    try:
        response = requests.post(url, headers=headers, data=payload, timeout=30)
        
        if response.status_code == 200:
            data = response.json()
            return data.get('access_token')
        else:
            print(f"❌ Ошибка получения токена: {response.status_code}")
            return None
            
    except requests.exceptions.RequestException as e:
        print(f"❌ Ошибка запроса токена: {e}")
        return None

def test_gigachat_chat():
    print("=== ТЕСТ ЧАТА С GIGACHAT (Python) ===")
    
    # Получаем токен
    token = get_gigachat_token()
    
    if not token:
        print("❌ Не удалось получить токен")
        return False
    
    print(f"✅ Токен получен: {token[:50]}...")
    
    # Тестируем запрос к моделям
    url = "https://gigachat.devices.sberbank.ru/api/v1/models"
    
    headers = {
        'Accept': 'application/json',
        'Authorization': f'Bearer {token}'
    }
    
    print("Тестируем запрос к моделям...")
    
    try:
        response = requests.get(url, headers=headers, timeout=30)
        
        print(f"✅ Статус ответа: {response.status_code}")
        print(f"Ответ: {response.text}")
        
        if response.status_code == 200:
            print("✅ Запрос к моделям успешен!")
            
            # Тестируем чат
            chat_url = "https://gigachat.devices.sberbank.ru/api/v1/chat/completions"
            
            payload = {
                "model": "GigaChat-2-Pro",
                "messages": [
                    {
                        "role": "user",
                        "content": "Привет! Ответь одним словом: \"Работает\""
                    }
                ],
                "temperature": 0.7,
                "max_tokens": 100
            }
            
            headers = {
                'Accept': 'application/json',
                'Authorization': f'Bearer {token}',
                'Content-Type': 'application/json'
            }
            
            print("Тестируем чат...")
            
            response = requests.post(chat_url, headers=headers, json=payload, timeout=30)
            
            print(f"✅ Статус ответа чата: {response.status_code}")
            print(f"Ответ чата: {response.text}")
            
            if response.status_code == 200:
                data = response.json()
                if 'choices' in data and len(data['choices']) > 0:
                    message = data['choices'][0].get('message', {}).get('content', '')
                    print(f"✅ Ответ GigaChat: {message}")
                    return True
                else:
                    print("❌ Неожиданный формат ответа чата")
                    return False
            else:
                print(f"❌ Ошибка чата: {response.status_code}")
                return False
        else:
            print(f"❌ Ошибка запроса к моделям: {response.status_code}")
            return False
            
    except requests.exceptions.RequestException as e:
        print(f"❌ Ошибка запроса: {e}")
        return False

if __name__ == "__main__":
    success = test_gigachat_chat()
    
    if success:
        print("🎉 Тест чата завершен успешно!")
    else:
        print("💥 Тест чата завершен с ошибкой!")

File: test-gigachat-chat.js
import axios from 'axios';
import https from 'https';
import fs from 'fs';
import path from 'path';
import { v4 as uuidv4 } from 'uuid';
import dotenv from 'dotenv';

// Загружаем переменные окружения
dotenv.config({ path: '.env.local' });

// Настройка сертификатов
const certPath = path.resolve(process.cwd(), 'russian_trusted_root_ca.crt');
if (fs.existsSync(certPath)) {
  process.env.NODE_EXTRA_CA_CERTS = certPath;
  console.log('✅ Сертификат НУЦ Минцифры загружен:', certPath);
} else {
  console.log('⚠️ Сертификат НУЦ Минцифры не найден, используем отключение проверки SSL');
  process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0';
}

async function getGigaChatToken() {
  try {
    const authKey = process.env.GIGACHAT_API_KEY;
    const rqUID = uuidv4();
    
    console.log('Получаем токен доступа...');
    
    const response = await axios.post('https://ngw.devices.sberbank.ru:9443/api/v2/oauth', 
      'scope=GIGACHAT_API_PERS',
      {
        headers: {
          'Authorization': `Basic ${authKey}`,
          'RqUID': rqUID,
          'Content-Type': 'application/x-www-form-urlencoded',
          'Accept': 'application/json'
        },
        httpsAgent: new https.Agent({
          rejectUnauthorized: false,
          secureProtocol: 'TLSv1_2_method',
          ca: fs.existsSync(certPath) ? fs.readFileSync(certPath) : undefined
        }),
        timeout: 30000,
        maxRedirects: 5
      }
    );
    
    console.log('✅ Токен получен успешно');
    return response.data.access_token;
  } catch (error) {
    console.error('❌ Ошибка получения токена:', error.message);
    throw error;
  }
}

async function testGigaChatChat() {
  try {
    console.log('=== ТЕСТ ЧАТА С GIGACHAT ===');
    
    // Получаем токен
    const accessToken = await getGigaChatToken();
    console.log('Access Token (first 50 chars):', accessToken.substring(0, 50) + '...');
    
    // Отправляем запрос к чату
    console.log('Отправляем запрос к GigaChat API...');
    console.log('URL: https://gigachat.devices.sberbank.ru/api/v1/chat/completions');
    
    const requestBody = {
      model: 'GigaChat-2-Pro',
      messages: [
        {
          role: 'user',
          content: 'Привет! Ответь одним словом: "Работает"'
        }
      ],
      temperature: 0.7,
      max_tokens: 100,
    };
    
    console.log('Request body:', JSON.stringify(requestBody, null, 2));
    
    const response = await axios.post('https://gigachat.devices.sberbank.ru/api/v1/chat/completions', 
      requestBody,
      {
        headers: {
          'Authorization': `Bearer ${accessToken}`,
          'Content-Type': 'application/json',
        },
        httpsAgent: new https.Agent({
          rejectUnauthorized: false,
          ca: fs.existsSync(certPath) ? fs.readFileSync(certPath) : undefined
        }),
        timeout: 30000
      }
    );
    
    console.log('✅ УСПЕХ! Ответ получен');
    console.log('Status:', response.status);
    console.log('Response data:', response.data);
    
    if (response.data.choices && response.data.choices[0] && response.data.choices[0].message) {
      console.log('Ответ GigaChat:', response.data.choices[0].message.content);
    }
    
    return response.data;
  } catch (error) {
    console.error('❌ ОШИБКА при отправке запроса к чату:');
    console.error('Error type:', error.constructor.name);
    console.error('Error message:', error.message);
    
    if (error.response) {
      console.error('Response status:', error.response.status);
      console.error('Response headers:', error.response.headers);
      console.error('Response data:', error.response.data);
    } else if (error.request) {
      console.error('Request was made but no response received');
      console.error('Request details:', error.request);
    }
    
    console.error('Full error:', error);
    throw error;
  }
}

// Запускаем тест
testGigaChatChat()
  .then(result => {
    console.log('🎉 Тест чата завершен успешно!');
    process.exit(0);
  })
  .catch(error => {
    console.error('💥 Тест чата завершен с ошибкой!');
    process.exit(1);
  });

